File: my_project/loans.py
# This file contain loans related routes

from datetime import datetime
from flask import Blueprint, jsonify, request
from sqlalchemy import func, null
from my_project import db
from my_project.models import Book, Customer, Loan


loans = Blueprint('loans', __name__)


# Route to create a new loan
@loans.route('/loans', methods=['POST'])
def create_loan():
    data = request.get_json()
    
    # Validate that required fields are present in the request data
    if 'bookID' not in data or 'customerID' not in data:
        return jsonify({'error': 'Required fields are missing in the request data (must be: bookID, and customerID)'}, 400)
    
    bookID = data['bookID']
    customerID = data['customerID']
    loanDate = datetime.now() # set the loan date to the current date

    
    # Check if the book with the specified bookID exists
    book = Book.query.get(bookID)

    # Check if the customer with the specified customerID exists
    customer = Customer.query.get(customerID)
    
    if book and book.status == 'available' and customer:
        # Update the book status to "unavailable"
        book.status = 'unavailable'

        # Create a new loan
        new_loan = Loan(loanDate=loanDate, bookID=bookID, customerID=customerID)
        db.session.add(new_loan)

        db.session.commit()

        return jsonify({'message': 'Loan created successfully!'})
    elif not customer:
        return jsonify({'error': 'Customer not found'}, 404)
    elif book and book.status == 'unavailable':
        return jsonify({'error': 'Book is already loaned'}, 400)
    else:
        return jsonify({'error': 'Book not found'}, 404)

# Route to retrieve all loans
@loans.route('/loans', methods=['GET'])
def get_loans():
    loans = Loan.query.all()
    
    loan_list = [{'loanID': loan.loanID, 'loanDate': loan.loanDate, 'MaxReturnDate': loan.maxReturnDate, 'ReturnDate': loan.returnDate, 'bookTitle': loan.book.title, 'customerName': loan.customer.Name} for loan in loans]

    return jsonify({'loans': loan_list})

# Route to end a loan
@loans.route('/loans/<int:loan_id>', methods=['PUT'])
def end_loan(loan_id):
    # Check if there is an active loan with the provided loan ID
    loan = Loan.query.filter_by(loanID=loan_id, returnDate=None).first()
    if loan:
        # Set the returnDate to the current date
        loan.returnDate = datetime.now()

        #check if the book exists, and update the book status to 'available'
        book = Book.query.get(loan.bookID)
        if book:
            book.status = 'available'

        db.session.commit()

        return jsonify({'message': 'Loan ended successfully'})
    else:
        return jsonify({'error': 'No active loan found with the provided loan ID'}, 400)

# Route to retrieve loans for which the books haven't been returned despite the expiration of the maximum return date.
@loans.route('/loans/late', methods=['GET'])
def get_late_loans():
    current_date = datetime.now().date()

    # filter late and active loans
    late_loans = Loan.query.join(Book, Loan.bookID == Book.bookID).filter(
        Loan.maxReturnDate < current_date,
        Loan.returnDate.is_(None)
    ).all()

    late_loan_list = [{'loanID': loan.loanID, 'loanDate': loan.loanDate, 'MaxReturnDate': loan.maxReturnDate, 'ReturnDate': loan.returnDate, 'bookTitle': loan.book.title, 'customerName': loan.customer.Name} for loan in late_loans]

    return jsonify({'late_loans': late_loan_list})

File: my_project/templates/static/js/loans.js
// This file conatain JS for "loans.html" file

let allLoans = []
let activeLoans = []
let lateLoans = []



// Create table headers
function createTableHeaders() {
    const tableHeader = document.createElement('thead');
    tableHeader.innerHTML = `
        <tr>
            <th>Book</th>
            <th>Customer</th>
            <th>Loan date</th>
            <th>Latest date to return</th>
            <th>Return date</th>
        </tr>
    `;
    loansTable.appendChild(tableHeader);
}

// Get an array and display it in the table body
function displayLoans(loans) {
    const tableBody = document.createElement('tbody');
    
    // Remove existing rows from the table
    const existingTableBody = loansTable.getElementsByTagName('tbody')[0];
    if (existingTableBody) existingTableBody.remove();

    // Use forEach loop to display loans
    loans.forEach(function (loan) {
        // Check if book was returned and returnDate exists, else return: "-"
        const ReturnDate = loan.ReturnDate ? formatDate(loan.ReturnDate) : "-";
        const row = document.createElement('tr');
        row.innerHTML = `
        <td>${loan.bookTitle}</td>
        <td>${loan.customerName}</td>
        <td>${formatDate(loan.loanDate)}</td>
        <td>${formatDate(loan.MaxReturnDate)}</td>
        <td>${ReturnDate}</td>
        <td>
            ${loan.ReturnDate ? '' : `<button onclick="returnBook(${loan.loanID})">Return book</button>`}
        </td>
    `;
        tableBody.appendChild(row);
    });
    document.getElementById('loansTable').appendChild(tableBody);
}

// Fetch and display all loans using Axios
function getAllLoans() {
    // send GET request to the Flask server
    axios.get('/loans')
        .then(function (response) {
            allLoans = response.data.loans; // Assign the fetched loans to the allLoans variable
            // Display search button
            searchDiv.innerHTML =   '<input id="searchInput" style="margin-left: 85px;" type="text" onkeypress="search(allLoans)" placeholder="Enter book to search"> <button onclick="getAllLoans()">Cancel</button>'
            // Set the currently active side bar item color
            b.style.backgroundColor = '#04AA6D'
            c.style.backgroundColor = '#04AA6D'
            a.style.backgroundColor = '#333'
            // Display all loans initially
            displayLoans(allLoans);
        })
        .catch(function (error) {
            console.error('Error fetching loans:', error);
        });
}

// Call createTableHeaders and getAllLoans once when the page loads
document.addEventListener('DOMContentLoaded', createTableHeaders);
document.addEventListener('DOMContentLoaded', getAllLoans);

// Display active loans only
function displayActiveLoans() {
    // Filter out loans with no returnDate to get only active loans
    activeLoans = allLoans.filter(function (loan) {
        return !loan.ReturnDate;
    });
    // Check if there are no active loans, and tostify it, else - display active loans
    if (activeLoans.length === 0) showSuccessNotification("There are no active loans.")
    else{
        // Display the relevant search button
        searchDiv.innerHTML =   '<input id="searchInput" style="margin-left: 85px;" type="text" onkeypress="search(activeLoans)" placeholder="Enter book to search"> <button onclick="displayActiveLoans()">Cancel</button>'
        // Set the currently active side bar item color
        a.style.backgroundColor = '#04AA6D'
        c.style.backgroundColor = '#04AA6D'
        b.style.backgroundColor = '#333'
        // Clear the search input
        searchInput.value = ''
        // display it in the table
        displayLoans(activeLoans);
    }
}

// Display late loans (loans for which the books haven't been returned despite the expiration of the maximum return date).
function displayLateLoans() {
    // send GET request to flask server to get late loans
    axios.get('/loans/late')
        .then(function (response) {
            lateLoans = response.data.late_loans;

            // Use the ternary operator to conditionally show a notification or display loans
            if(lateLoans.length === 0) showSuccessNotification("There are no late loans.")
            else{
                // Display the relevant search button
                searchDiv.innerHTML =   '<input id="searchInput" style="margin-left: 85px;" type="text" onkeypress="search(lateLoans)" placeholder="Enter book to search"> <button onclick="displayLateLoans()">Cancel</button>'
                // Set the currently active side bar item color
                b.style.backgroundColor = '#04AA6D'
                a.style.backgroundColor = '#04AA6D'
                c.style.backgroundColor = '#333'
                // Clear the search input field
                searchInput.value = ''
                // display the lateLoans array in the table
                displayLoans(lateLoans);
            }
        })
        .catch(function (error) {
            console.error('Error fetching late loans:', error);
            showErrorNotification("Error fetching late loans.");
        });
}

// Format a date string to a simpler format
function formatDate(dateString) {
    const options = { year: 'numeric', month: 'numeric', day: 'numeric' };
    return new Date(dateString).toLocaleDateString(undefined, options);
}

// Ruturn a book and end the loan
function returnBook(loanID) {
    // Ask for confirmation from the user
    const userConfirmed = confirm("Are you sure you want to return this book?");

    if (userConfirmed) {
        // Send a PUT request to the /loans/<loan_id> endpoint to end the loan
        axios.put(`/loans/${loanID}`)
            .then(function (response) {
                // Check if the request was successful
                if (response.data.message === 'Loan ended successfully') {
                    // Display a success notification
                    showSuccessNotification('Book returned successfully');
                    // Reload display active loans after ending the loan
                    getAllLoans();
                }
            })
            .catch(function (error) {
                console.error('Error returning book:', error);
                showErrorNotification('Error returning book');
            });
    } else {
        // The user canceled the action
        console.log('Book return action canceled by the user');
    }
}

// Get an array and search for a specific loan (by book title), using "filter"
function search(loans){
    const search_input = searchInput.value.toLowerCase();
    const filterLoans = loans.filter(function(loan){
        return loan.bookTitle.toLowerCase().includes(search_input)
    });
    // display the filtered array in the table
    displayLoans(filterLoans)
}